### DISCUSSION TITLE
Invisible Damaging Fire

### DISCUSSION THREAD

**Discussion by Mixmaster49** [FakeFishGames/Barotrauma#15355] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

Over multiple instances in the Submarine pomegrenade plantation room there have been invisible fires that deal insane amounts of burn damage and stun to anyone who enters the room. This damage is enough to stunlock and potentially kill a player wearing a PUCS. The room this occurred in was a diving room, with oxygen tanks connected to power but diving suits not, and a vent connected to the oxygen generator. 

This bug has been consistently occurring, but I am not sure what exactly is causing the fire to act this way. There are many more normal fires than invisible fires seen. 

### Reproduction steps

1. Start a game
2. Place a large number of planter boxes into a room and seed them with pomegrenade
3. Occasionally, this kind of invisible fire will occur

### Bug prevalence

Happens every now and then

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.7.7.0 (Winter Update)

### Which operating system did you encounter this bug on?

Windows

**Comment by SomeRandomNoobKekeke** [comment]
> invisible fires that deal insane amounts of burn damage and stun

Are those fires also throw people around? Obviously fires can't do that, that must be some stray explosion from some status effect

I've seen this before, in my case there was a fire near oxygen tank shelf, and it somehow created a zone where you instantly die from burns and get thrown away

I was never able to reproduce, so i never reported it, and iirc there was performance fix in that session, do you also have it?

**Reply by Mixmaster49** [reply]
Yeah it was this sorta thing. It might have something to do with the oxygen shelves because it was in our diving room.

**Reply by Regalis11** [reply]
Are you using the Performance Fix mod? I think it could potentially cause these sorts of issues due to the way it forces some things to only update occasionally.

**Reply by Mixmaster49** [reply]
No mods besides workshop and custom submarines are being used.

**Comment by SomeRandomNoobKekeke** [comment]
Well, it's possible to trigger endless chain explosion of oxygen tanks without performance fix
Perhaps some lucky arrangement could create more "invisible wall" like effect

https://github.com/user-attachments/assets/5936360e-76d3-44e9-824c-ee73c942121f

**Comment by SomeRandomNoobKekeke** [comment]
Accidentally created this anomaly on azimuth. As you can see full vanilla

https://github.com/user-attachments/assets/17274f33-372f-48cc-85ea-edeab07e715c

**Reply by ubertpendragon** [reply]
Not really related to invisible fires: 

It's just that explosions have 1 tick of "fire" damage hardcoded in (which causes problems)
Shelves will not refill O2 tanks if they have the "onfire" tag. Tanks get the "onfire" tag for 1 second upon taking fire damage.
That tag only sticks if there's a constant fire source nearby, but periodic explosions don't do that

It's silly that O2 tanks do not EXPLODE when they EXPLODE, solving this feedback loop once and for all (and giving a sink for infinite aluminium piles)

**Reply by Regalis11** [reply]
> It's just that explosions have 1 tick of "fire" damage hardcoded in (which causes problems)
> Shelves will not refill O2 tanks if they have the "onfire" tag. Tanks get the "onfire" tag for 1 second upon taking fire damage.
> That tag only sticks if there's a constant fire source nearby, but periodic explosions don't do that

I don't quite follow, what do you mean the tag only sticks if there's a constant fire source nearby? It has a duration of 1 second, it doesn't require the a constant fire source.

**Reply by ubertpendragon** [reply]
The sequence of events goes like this:

1. OnFire reduces condition of O2 tank -10/s and adds "onfire" tag to the tank for a duration of 1 second

2. When O2 tank reaches 0 condition and has "onfire" tag, it explodes after a 1 second delay

3. Oxygen shelves refill O2 tanks UNLESS they have "onfire" tag

So the problem is the following:
If tank1 explodes, and tank2 has <10 condition, tank2 is tagged by the explosion of tank1 "onfire" for 1 second then explodes after 1s . Then Tank1 loses its "onfire" tag after 1s, gains condition from the O2 shelves, gets tagged by the explosion of tank2...
Both tanks will explode in an unending loop until you turn off power.

The point is that **there is no invisible fire**, it's just a loop of O2 tanks _pseudo_-exploding (pseudo because the tank can be reused) triggering other nearby O2 tanks

**Comment by ubertpendragon** [comment]
https://github.com/user-attachments/assets/cdef746d-0fa1-478a-9171-1328e6c0e95d

**Reply by Regalis11** [reply]
I'm a little confused by this video, is this an example of the invisible fire? To me it seems like it's normal, visible explosions, which are for some reason triggering continuously. Do you happen to have any idea how to get this to occur?

**Reply by SomeRandomNoobKekeke** [reply]
It takes some luck to reproduce. It's Azimuth btw

https://github.com/user-attachments/assets/babc49cf-dfa0-4e41-938c-b7d92b477193

Not my report, but when i first saw this i thought it's fire because it was right where i just extinguished it and it dealt burning damage

**Comment by JustKesha** [comment]
I've had invisible fires happen multiple times during my recent playthroughs, tho not from plants
Entire floor of the sub will be burning and emitting light but no fire effect will be seen

**Reply by SomeRandomNoobKekeke** [reply]
That's probably just imaginary fire from psychosis turned into real fire
![Screenshot_3](https://github.com/user-attachments/assets/faa21d18-e16d-458b-b012-de8d09cd0795)
I can't find the ticket, i guess it was miraculously fixed before anyone reported it

Anyway this report is about chain exposion of o2 tanks that looks like invisible fire but isn't

**Comment by Regalis11** [comment]
I believe this has now been addressed in our private repo (https://github.com/FakeFishGames/Barotrauma-development/pull/6727), fix coming up in the next unstable build.

There were a couple of issues here that I'm fairly sure were causing the bug:
- The "explosion loops" caused by oxygen tank shelves
- Clients did not see the tank explosions at their end, so it would seem like they're taking burn damage out of nowhere

So I think this was most likely actually a case of invisible explosions, not invisible fire.
